hipamd/src/hip_mrfs.cpp

```diff
@@ -24,208 +28,610 @@
 
 namespace hip {
 
+// Default to 1GB chunk size
+constexpr size_t DEFAULT_MEMCPY_CHUNK_SIZE = 128ULL * 1024 * 1024;
+static size_t gMemcpyChunkSize = DEFAULT_MEMCPY_CHUNK_SIZE;
+
+// Default process bandwidth quota (10 GB/s)
+constexpr double DEFAULT_PROCESS_BANDWIDTH_QUOTA = 20.0 * 1024 * 1024 * 1024;
+
+size_t getMemcpyChunkSize() { return gMemcpyChunkSize; }
+
+void setMemcpyChunkSize(size_t newChunkSize) {
+  if (newChunkSize > 0) {
+    gMemcpyChunkSize = newChunkSize;
+  }
+}
+
 // Forward declarations
 class Device;
 class Stream;
-extern hipError_t ihipDeviceSynchronize();
-extern hipError_t ihipStreamSynchronize(hipStream_t stream);
 extern hipError_t ihipMemcpy_validate(void* dst, const void* src, size_t sizeBytes,
                                       hipMemcpyKind kind);
-// extern hipError_t ihipMemcpy(void* dst, const void* src, size_t sizeBytes,
-//                            hipMemcpyKind kind, hip::Stream& stream, bool isAsync);
 extern Device* getCurrentDevice();
 
+struct QueueId {
+  int deviceId;
+  pid_t processId;
+
+  // For using QueueId as a map key
+  bool operator<(const QueueId& other) const {
+    if (deviceId != other.deviceId) return deviceId < other.deviceId;
+    return processId < other.processId;
+  }
+
+  bool operator==(const QueueId& other) const {
+    return deviceId == other.deviceId && processId == other.processId;
+  }
+};
+
 // Struct used to keep track of memory copy operations
 struct MemcpyTask {
   void* dst;
   const void* src;
   size_t sizeBytes;
   hipMemcpyKind kind;
   hip::Stream* stream;
-  int deviceId;
+  QueueId queueId;
   std::atomic<bool> completed{false};
 };
 
-// Thread pool for paralell memory operations
+// quota management
+class QuotaManager {
+  private:
+  // Store bandwidth quotas in bytes per second for each (device, process) queue
+  std::map<QueueId, double> _queueQuotas;
+  std::map<QueueId, double> _queueUsage;
+  std::mutex _mutex;
+  std::thread _resetThread;
+  std::atomic<bool> _stopResetThread{false};
+  double _totalProcessQuota;
+  std::map<pid_t, double> _processQuotas;
+  std::map<pid_t, double> _processUsage;  // Track usage at process level
+
+  void quotaResetThreadFunction() {
+    MRFS_LOG("QUOTA", "Quota reset thread started");
+    while (!_stopResetThread) {
+      // Sleep for 1 second
+      struct timespec ts;
+      ts.tv_sec = 1;  // 1 second
```

Comment:
Sleep for shorter time and decrement usage by a fraction of the total bandwidth instead.